### PR TITLE
crypto in core/token network.update_fee 

### PR DESCRIPTION
**Pull request by err508**
fee type handling in update_fee is not beautiful, but working

### CODE REVIEW

**Comment by LefterisJP** [comment]
@err508 lots of flake8 failures. Integrate flake8 with your editor so that you get real-time indications of style as you are coding.

File: requirements.txt [hunk 1]
```diff
@@ -12,6 +12,7 @@ flask_restful
 gevent
 networkx
 requests
+numpy
```
Comment (LefterisJP) [inline requirements.txt]: shouldn't this be only inside requirements_dev ?

**Comment by devium** [comment]
As discussed, I think we should merge #16 before this one because that enables this PR to get rid of the procedurally generated participants and deposits, getting rid of the multiple `random.seed()` calls.

Channel participants and deposits would then already be known by the `TokenNetwork` through initial calls to `handle_channel_open_event` and `handle_channel_new_deposit_event` without the call to the patched getters in `TokenNetworkContract`. After rebasing, this PR would use @palango's method of initially generating a random network in #16 instead of (re)generating participants and deposits on the fly.

File: pathfinder/token_network.py [hunk 2]
```diff
@@ -146,25 +148,29 @@ def update_balance(
...
+                msg,
+                hasher=sha256)
+        )
         participant1, participant2 = self.token_network_contract.get_channel_participants(
             channel_id
         )
         if is_same_address(participant1, signer):
             sender = participant1
             receiver = participant2
+            new_fee = float(new_fee)
```
Comment (palango) [inline pathfinder/token_network.py]: Shouldn't these be `int`s? We should probably define if we handle values in the smallest denomination or relative to a "full" token itself . @devium

File: requirements_dev.txt [hunk 3]
```diff
@@ -11,7 +11,7 @@ sphinx_rtd_theme==0.2.4
 
 pytest==3.4.2
 pytest-runner==4.0
-
+numpy
```
Comment (palango) [inline requirements_dev.txt]: I already added this. Can be removed.

File: requirements.txt [hunk 4]
```diff
@@ -13,5 +13,6 @@ gevent
 networkx
 requests
 
+
```
Comment (palango) [inline requirements.txt]: Please remove.

File: pathfinder/token_network.py [hunk 5]
```diff
@@ -184,21 +190,32 @@ def update_fee(
...
             view: ChannelView = attr['view']
             assert 0 <= hop_bias <= 1
             if view.capacity < value:
                 return None
             else:
-                return hop_bias*self.max_fee + (1-hop_bias)*view.fee + visited.get(
-                    view.channel_id,
-                    0
-                )
+                return hop_bias*self.max_fee + \
```
Comment (palango) [inline pathfinder/token_network.py]: Please put spaces around the operators

File: requirements_dev.txt (hunk 3 above)
Comment (LefterisJP) [reply requirements_dev.txt]: It will be removed/ignored at rebase time :)

File: pathfinder/tests/fixtures/accounts.py [hunk 6]
```diff
@@ -22,24 +23,24 @@ def faucet_address(faucet_private_key: str):
 
 
 @pytest.fixture(scope='session')
-def private_keys() -> List[str]:
+def private_key_seeds() -> List[str]:
```
Comment (devium) [inline pathfinder/tests/fixtures/accounts.py]: Why did you rename this? It's still a list of actual Ethereum private keys (hex-encoded). I get your critique that this isn't the ECDSA private key but still, we call this the private key in Ethereum. See the yellow paper if you don't believe me.

File: pathfinder/tests/fixtures/core.py [hunk 7]
```diff
@@ -1,42 +1,47 @@
...
 
 @pytest.fixture
 def populate_token_networks(
         token_networks: List[TokenNetwork],
         token_network_contracts: List[Contract],
+        addresses: List[Address],
+        private_key_seeds: List[str],
 ) -> None:
-
+    random.seed(NUMBER_OF_CHANNELS)
```
Comment (devium) [inline pathfinder/tests/fixtures/core.py]: Explicit seed is good for reproducible testing but why `NUMBER_OF_CHANNELS`? `0` would be just fine too.

File: pathfinder/tests/fixtures/core.py [hunk 8]
```diff
@@ -1,42 +1,47 @@
...
-
-            p1 = to_checksum_address(encode_hex(keccak(seed1))[:42])
-            p2 = to_checksum_address(encode_hex(keccak(seed2))[:42])
-
+        for channel_id in range(NUMBER_OF_CHANNELS):
+            seed1, seed2 = random.sample(private_key_seeds, 2)
+            private_key1 = PrivateKey.from_hex(remove_0x_prefix(seed1))
+            private_key2 = PrivateKey.from_hex(remove_0x_prefix(seed2))
+            public_key1 = to_checksum_address(public_key_to_address(private_key1.public_key))
+            public_key2 = to_checksum_address(public_key_to_address(private_key2.public_key))
```
Comment (devium) [inline pathfinder/tests/fixtures/core.py]: These are addresses, not public keys. Ethereum doesn't have a different definition of public key as it does with its private key counterpart. An Ethereum public key is the ECDSA public key, which again is different from the address. You cannot even reconstruct the public key from an address. See the definition of `public_key_to_address` for more information.

Renaming suggestions here:
* `seed` to `private_key`
* `public_key` to `address`
* `private_key` to `private_key_ecdsa` or something, to differentiate it from the Ethereum private key

File: pathfinder/token_network.py [hunk 9]
```diff
@@ -155,27 +155,28 @@ def update_balance(
...
-        # signer = from_signature_and_message(signature, msg)
-        # <-- this verifies the signature of the fee-update!
-        signer = signature
+        # Fixme: I need a nonce for replay protection
+        msg = new_fee
+        signer = public_key_to_address(
+            PublicKey.from_signature_and_message(
+                signature,
+                msg,
+                hasher=sha256)
```
Comment (devium) [inline pathfinder/token_network.py]: Closing parentheses on new line.

File: pathfinder/token_network.py [hunk 10]
```diff
@@ -155,27 +155,28 @@ def update_balance(
         )
 
     def update_fee(
-        self,
-        channel_id: ChannelId,
-        new_fee: float,
-        signature: Address
+            self,
+            channel_id: ChannelId,
+            new_fee: bytes,
```
Comment (devium) [inline pathfinder/token_network.py]: I get that you probably don't want `new_fee` to be `float` if you need it as a signed message, but I guess using `str` here and `new_fee` being something like `'0.00058'` makes the most sense. It avoids the precision issue as well as the unique floating point value representation issue. I'm open to discussion on this one though, as there might be several solutions. I just don't think `bytes` is the prettiest.

Client-side it's also a lot easier to convert `float` into `str` than into `bytes`. But that's just my two cents.

File: pathfinder/token_network.py [hunk 11]
```diff
@@ -155,27 +155,28 @@ def update_balance(
...
+        )
         participant1, participant2 = self.channel_id_to_addresses[channel_id]
         if is_same_address(participant1, signer):
             sender = participant1
             receiver = participant2
+            new_fee = float(new_fee)
         elif is_same_address(participant2, signer):
             sender = participant2
             receiver = participant1
+            new_fee = float(new_fee)
```
Comment (devium) [inline pathfinder/token_network.py]: Does this work on a `bytes` object? If we stick with `bytes` please specify the internal representation of `new_fee` in a docstring.

File: pathfinder/token_network.py [hunk 12]
```diff
@@ -196,18 +197,22 @@ def update_fee(
...
         hop_bias = kwargs.get('bias', 0)
 
-        def weight(u: Address, v: Address, attr: Dict[str, Any]):
+        def weight(
+                u: Address,
+                v: Address,
+                attr: Dict[str, Any]
+        ):
             view: ChannelView = attr['view']
             assert 0 <= hop_bias <= 1
```
Comment (devium) [inline pathfinder/token_network.py]: Move the `hop_bias` check out of the `weight` function. It's constant for a single `get_paths` call so we don't need to check it on each edge traversal.

File: pathfinder/tests/fixtures/core.py [hunk 13]
```diff
@@ -1,42 +1,47 @@
...
+                public_key2
             )
 
-        # deposit to channels
-        for channel_id in range(100):
-            deposit1 = random.randint(0, 1000)
-            deposit2 = random.randint(0, 1000)
-
+            # deposit to channels
+            deposit1, deposit2 = random.sample(range(1000), 2)
```
Comment (devium) [inline pathfinder/tests/fixtures/core.py]: We might want to make the max deposit configurable, similar to the `NUMBER_OF_CHANNELS` and `NUMBER_OF_NODES`. Works for now though.

**Comment by devium** [comment]
On a general note: I know I named the fixture file `core.py` but we should probably rename that. Maybe `network.py` or `service.py`

File: pathfinder/tests/fixtures/core.py (hunk 7 above)
Comment (err508) [reply pathfinder/tests/fixtures/core.py]: to make it look cool!

File: pathfinder/tests/fixtures/accounts.py (hunk 6 above)
Comment (err508) [reply pathfinder/tests/fixtures/accounts.py]: No offence, it's just a list of hashes of an ascending sequence of natural numbers, encoded as hex. Calling it private keys seems dangerous. I think even calling the result of privatekey.from_hex() on it a private key is dangerous. At some point someone might start using this, we should make it hard to make mistakes.

File: pathfinder/tests/fixtures/core.py (hunk 8 above)
Comment (err508) [reply pathfinder/tests/fixtures/core.py]: Agreed on 2 and 3!

File: pathfinder/tests/fixtures/accounts.py (hunk 6 above)
Comment (devium) [reply pathfinder/tests/fixtures/accounts.py]: They are private keys because of what they are used for, not because of how they were created. We're using deterministically generated private keys. Nothing wrong with that in a test setup.

There are secure and insecure private keys. These are definitely the latter but they are still private keys.

File: pathfinder/tests/fixtures/core.py (hunk 7 above)
Comment (devium) [reply pathfinder/tests/fixtures/core.py]: It's confusing. People reading this will wonder why the seed is exactly `NUMBER_OF_CHANNELS` so please change that or document that you just use that variable to make it look cool according to your standards of cool.

File: pathfinder/tests/fixtures/accounts.py (hunk 6 above)
Comment (devium) [reply pathfinder/tests/fixtures/accounts.py]: This still holds after the recent push. Please rename to `private_key` because that's what it is and how it is used.

File: pathfinder/tests/fixtures/core.py (hunk 7 above)
Comment (err508) [reply pathfinder/tests/fixtures/core.py]: Added a comment.

File: pathfinder/token_network.py (hunk 10 above)
Comment (palango) [reply pathfinder/token_network.py]: @devium Shouldn't we make this an `int` and the unit the token wad (apparently the wei for tokens)

Comment (err508) [reply pathfinder/token_network.py]: Bytes are necessary for proper hashing and signature verification. Bytes are constructed from string via str().encode, so no float problems here. Float is only used for internal computation and stored in channel_view.

Comment (pcppcp) [reply pathfinder/token_network.py]: you should probably use `struct` if you are dealing with binary data.
https://github.com/raiden-network/raiden-libs/blob/master/raiden_libs/messages/balance_proof.py#L58

Comment (pcppcp) [reply pathfinder/token_network.py]: regarding use of float, it's always better to use (fixed point) integer.

File: pathfinder/tests/fixtures/accounts.py (hunk 6 above)
Comment (err508) [reply pathfinder/tests/fixtures/accounts.py]: Ok, done.

File: pathfinder/tests/fixtures/core.py (hunk 7 above)
Comment (pcppcp) [reply pathfinder/tests/fixtures/core.py]: don't use fixed seed for regular testing - running the test every time with set of variables derived from different seed values reveals more potential problems.

File: pathfinder/token_network.py (hunk 10 above)
Comment (devium) [reply pathfinder/token_network.py]: @palango Depends on the fee model we decide to go with. Fractions totally make sense over absolute token values. Fixed point makes sense of course but for calculations it really doesn't matter. `float` precision is likely good enough for routing. So no issue with float vs fixed point but we need a deterministic representation for signature verification. `str` works in that regard (or the `bytes` representation thereof).

File: pathfinder/tests/fixtures/core.py (hunk 7 above)
Comment (devium) [reply pathfinder/tests/fixtures/core.py]: @pcppcp These `TokenNetwork`s are intended for unit testing. We need those networks to be deterministic for reproducibility. Some networks might be partitioned in which case tests would probably fail, so it's good to have one working network to run tests on.

Alternatively we could use a real random seed and log it each time for reproducibility, so we can then debug against a fixed seed. But still, having tests randomly fail and pass doesn't seem like a good idea to me.

Comment (pcppcp) [reply pathfinder/tests/fixtures/core.py]: Have you considered generating the network with `networkx` library? It can create various topologies that are guaranteed to be connected. Or re-generate network until you get connected graph.
Logging the seed is a good idea. You could set envvar/option that's set to i.e. `date +%s` from Travis or use python's `time` as a fallback.

File: pathfinder/token_network.py [hunk 14]
```diff
@@ -155,20 +155,23 @@ def update_balance(
...
-        signature: Address
+            self,
+            channel_id: ChannelId,
+            new_fee: bytes,
+            signature: bytes
     ):
-        """ Update the channel with a new fee.
-
-        Called by the public interface. """
+        """new_fee bytes are of the form '0.0012'.encode()"""
```
Comment (palango) [inline pathfinder/token_network.py]: Can we keep the first part of the docstring.
